src/useEntryGroup.ts

```diff
@@ -24,92 +24,133 @@ type Row =
+
+    if (this._leftChild === null) {
+      const { left, mid } = this;
+      this._leftChild = new SegmentTreeNode(0, [left, mid]);
+    }
+
+    return this._leftChild;
+  }
+
+  public get rightChild() {
+    if (this.isLeaf) {
+      throw new Error("Try accessing the right child of a leaf node.");
+    }
+
+    if (this._rightChild === null) {
+      const { right, mid } = this;
+      this._rightChild = new SegmentTreeNode(0, [mid, right]);
+    }
+
+    return this._rightChild;
   }
 }
 
 export class SegmentTree {
   private root: SegmentTreeNode;
-  private capacity: number;
+  private get capacity() {
+    const { right, left } = this.root;
+    return right - left;
+  }
   private length: number;
 
   constructor() {
-    this.root = new SegmentTreeNode();
-    this.capacity = 1;
+    this.root = new SegmentTreeNode(0, [0, 1]);
     this.length = 0;
   }
 
-  static createEmptyTree(depth: number): null | SegmentTreeNode {
-    if (depth === 0) {
-      return null;
-    }
-
-    return new SegmentTreeNode(
-      0,
-      SegmentTree.createEmptyTree(depth - 1),
-      SegmentTree.createEmptyTree(depth - 1)
-    );
-  }
-
   static updateIndex(
     node: SegmentTreeNode,
-    capacity: number,
     index: number,
-    value: number
+    value: number,
   ): void {
-    if (capacity === 1) {
-      if (node.left !== null && node.right !== null) {
-        throw new Error(
-          `can only update the value of leaf node, not the value of intermediate segment node`
-        );
-      }
+    if (node.isLeaf) {
       node.value = value;
       return;
     }
 
-    if (node.left === null || node.right === null) {
-      throw new Error(
-        `intermediate segment node must have left and right children`
-      );
+    const { left, right, mid } = node;
+
+    if (left <= index && index < mid) {
+      SegmentTree.updateIndex(node.leftChild, index, value);
     }
 
-    if (index < capacity / 2) {
-      SegmentTree.updateIndex(node.left, capacity / 2, index, value);
-    } else {
-      SegmentTree.updateIndex(
-        node.right,
-        capacity / 2,
-        index - capacity / 2,
-        value
-      );
+    if (mid <= index && index < right) {
+      SegmentTree.updateIndex(node.rightChild, index, value);
     }
 
-    node.value = node.left.value + node.right.value;
+    node.value = node.leftChild.value + node.rightChild.value;
   }
 
   /**
    * allocate makes its capacity twice the original capacity
    */
   private allocate(): void {
-    // make an empty tree in the same capacity
-    const depth = Math.log2(this.capacity);
-    const emptyTree = SegmentTree.createEmptyTree(depth + 1);
-    this.root = new SegmentTreeNode(this.root.value, this.root, emptyTree);
-    this.capacity *= 2;
+    const leftChild = this.root;
+    const capacity = 2 * this.capacity;
+    this.root = new SegmentTreeNode(leftChild.value, [0, capacity], leftChild);
   }
 
-  // time complexity: amortized O(log n)
+  // time complexity: O(1)
```

Comment:
Why can this operation be optimize to O(1) while still calling an O(log n) operation `set()` in line 158